File: backend/app/schemas/code_analysis.py
from datetime import datetime
from typing import Any, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class AnalyzeRepositoryRequest(BaseModel):
    path: Optional[str] = Field(default=None, description="Local filesystem path of repository source directory")


class CodeFileResponse(BaseModel):
    id: str
    repository_id: str
    path: str
    language: str
    size: int
    analysis_status: str
    analysis_error: Optional[str] = None
    analyzed_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CodeEntityResponse(BaseModel):
    id: str
    repository_id: str
    file_id: str
    parent_entity_id: Optional[str] = None
    entity_type: str  # FILE, MODULE, FUNCTION, CLASS, METHOD
    name: str
    qualified_name: Optional[str] = None
    start_line: int
    end_line: int

    model_config = ConfigDict(from_attributes=True)


class CodeRelationshipResponse(BaseModel):
    id: str
    repository_id: str
    source_entity_id: str
    target_entity_id: Optional[str] = None
    relationship_type: str  # CONTAINS, IMPORTS, CALLS, DEFINES
    resolved: bool
    raw_target: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class AnalysisRunResponse(BaseModel):
    id: str
    repository_id: str
    status: str
    files_discovered: int
    files_analyzed: int
    entities_found: int
    relationships_found: int
    files_failed: int
    started_at: datetime
    completed_at: Optional[datetime] = None
    error_summary: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class GraphNode(BaseModel):
    id: str
    label: str
    type: str  # repository, file, module, class, function, method
    file_id: Optional[str] = None
    file_path: Optional[str] = None
    name: str
    qualified_name: Optional[str] = None
    start_line: Optional[int] = None
    end_line: Optional[int] = None


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    relationship_type: str  # CONTAINS, IMPORTS, CALLS, DEFINES
    resolved: bool
    raw_target: Optional[str] = None


class GraphResponse(BaseModel):
    nodes: List[GraphNode]
    edges: List[GraphEdge]

File: frontend/src/types/code_analysis.ts
export interface CodeFile {
  id: string;
  repository_id: string;
  path: string;
  language: string;
  size: number;
  analysis_status: 'success' | 'error' | 'pending';
  analysis_error?: string | null;
  analyzed_at: string;
}

export interface CodeEntity {
  id: string;
  repository_id: string;
  file_id: string;
  parent_entity_id?: string | null;
  entity_type: 'FILE' | 'MODULE' | 'CLASS' | 'FUNCTION' | 'METHOD';
  name: string;
  qualified_name?: string | null;
  start_line: number;
  end_line: number;
}

export interface CodeRelationship {
  id: string;
  repository_id: string;
  source_entity_id: string;
  target_entity_id?: string | null;
  relationship_type: 'CONTAINS' | 'IMPORTS' | 'CALLS' | 'DEFINES';
  resolved: boolean;
  raw_target?: string | null;
}

export interface AnalysisRun {
  id: string;
  repository_id: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'partial';
  files_discovered: number;
  files_analyzed: number;
  entities_found: number;
  relationships_found: number;
  files_failed: number;
  started_at: string;
  completed_at?: string | null;
  error_summary?: string | null;
}

export interface GraphNode {
  id: string;
  label: string;
  type: string;
  file_id?: string | null;
  file_path?: string | null;
  name: string;
  qualified_name?: string | null;
  start_line?: number | null;
  end_line?: number | null;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  relationship_type: string;
  resolved: boolean;
  raw_target?: string | null;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}
